webnn/resources/utils.js

```diff
@@ -34,34 +34,59 @@ const loadTestData = (file) => {
     } else {
       throw new Error(`Failed to load ${file}`);
     }
-  }
+  };
 
-  const json = loadJSON(file);
+  const json = loadJSON();
   return JSON.parse(json.replace(/\\"|"(?:\\"|[^"])*"|(\/\/.*|\/\*[\s\S]*?\*\/)/g, (m, g) => g ? "" : m));
-}
+};
+
+/**
+ * Get exptected data source from given resources with output name.
+ * @param {Array} resources - An array of expected resources
+ * @param {String} name - An output name
+ * @returns {Number[]} An expected data array
+ */
+const getExpectedData = (resources, name) => {
+  let data;
+  for (let subResources of resources) {
+    if (subResources.name === name) {
+      data = subResources.data;
+      break;
+    }
+  }
+  if (data === undefined) {
+    throw new Error(`Failed to get expected data sources by ${name}`);
+  }
+  return data;
+};
 
 /**
  * Get ULP tolerance of gemm operation.
  * @param {Object} resources - Resources used for building a graph
- * @param {String} [opName] - An operation name
+ * @param {String} [name] - An operation name
  * @returns {Number} A tolerance number
  */
-const getGemmPrecisionTolerance = (resources, opName) => {
+const getGemmPrecisionTolerance = (resources, name) => {
   // GEMM : alpha * (A x B) + beta * C
   // An upper bound for the worst serial ordering is bounded by
   // the number of lossy operations, where matrix multiplication
   // is a dot product (mul and add times the number of elements)
   // plus bias operations.
-  const shapeA = resources.inputA.shape;
-  const options = resources.options;
+  const shapeA = resources.inputs['a'].shape;
+  const defaultOptions = {
+    c: 0.0,
+    alpha: 1.0,
+    beta: 1.0,
+    aTranspose: false,
+    bTranspose: false
+  }
+  const options = resources.options ? resources.options : defaultOptions;
```

Comment:
The default options are defined by spec and should be implemented by browser. 
spec: https://www.w3.org/TR/webnn/#dictdef-mlgemmoptions
chromium implementation: https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/ml/webnn/ml_graph_builder.idl;l=27

Reply:
What's the use case of getting the default settings? If user code doesn't set the value of an option, the default value will be used if it is defined by the spec.

Reply:
If you have concerns, I'd recommend to leave this function to following PRs. It would be good to have only code needed for this concat test case.

Reply:
OK, I will move it, and add it in next PR, thanks.